### PR TITLE
Add support for density compensation

### PR DESCRIPTION
**Pull request by chaithyagr**
I have made a quick pypi release of gpuNUFFT with support for density estimation. 

This PR mainly handles all the codes to have density compensation based reconstructions.

Starting the review as it is pretty much complete from my side (except any upcoming issues). This is the final PR before an official release of pysap-mri !!

Adding @Daval-G for testing the codes, please follow the methods and let me know if you face any issues.
Adding @zaccharieramzi in meantime for code and doc review.

### CODE REVIEW

**Comment by chaithyagr** [comment]
Rebased to `master` !

**Comment by chaithyagr** [comment]
> The code is working on my side, and the examples produce relevant results.
> However I noticed some old warnings:
> `UserWarning: pynufft python package has not been found. If needed use the master release. Till then you cannot use NUFFT on GPU`
> 
> Since this PR is related to the stable release of gpuNUFFT as a python package, it would be relevant to update related warnings in the code (pynufft is not the only GPU option proposed anymore). Also more formality in the message would be appreciated.

Yes, you are right, this might confuse the users. I will move to having the codes actually add this as an error only if the class is called, this is obsolete at this point.

**Comment by chaithyagr** [comment]
Will merge once @zaccharieramzi completes the final review...

File: examples/GPU_Examples/NonCartesian_gpuNUFFT_DensityCompensation.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
+"""
+Neuroimaging non-cartesian reconstruction
+=========================================
+
+Author: Chaithya G R
+
+In this tutorial we will reconstruct an MRI directly with density compensation
```
Comment (zaccharieramzi) [inline examples/GPU_Examples/NonCartesian_gpuNUFFT_DensityCompensation.py:7]: MRI image or MR image

File: examples/GPU_Examples/NonCartesian_gpuNUFFT_DensityCompensation.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+Import neuroimaging data
+------------------------
+
+We use the toy datasets available in pysap, more specifically a 2D brain slice
+and the acquisition radial non-cartesian scheme.
```
Comment (zaccharieramzi) [inline examples/GPU_Examples/NonCartesian_gpuNUFFT_DensityCompensation.py:13]: the radial acquisition scheme (non-cartesian), I think is better suited

File: examples/GPU_Examples/NonCartesian_gpuNUFFT_DensityCompensation.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+# Third party import
+from modopt.math.metrics import ssim
+from modopt.opt.linear import Identity
+from modopt.opt.proximity import SparseThreshold
+import numpy as np
+
+# Loading input data
+image = get_sample_data('2d-mri')
+
+# Obtain MRI non-cartesian mask
```
Comment (zaccharieramzi) [inline examples/GPU_Examples/NonCartesian_gpuNUFFT_DensityCompensation.py:34]: and estimate associated density compensation

File: examples/GPU_Examples/NonCartesian_gpuNUFFT_DensityCompensation.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    shape=image.shape,
+    implementation='gpuNUFFT',
+)
+fourier_op_density_comp = NonCartesianFFT(
+    samples=kspace_loc,
+    shape=image.shape,
+    implementation='gpuNUFFT',
+    density_comp=density_comp
+)
+kspace_obs = fourier_op.op(image.data)
```
Comment (zaccharieramzi) [inline examples/GPU_Examples/NonCartesian_gpuNUFFT_DensityCompensation.py:59]: specify that it would be the same with the density compensated op since the forward operation is the same

File: examples/GPU_Examples/NonCartesian_gpuNUFFT_DensityCompensation.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+# Simple adjoint
+image_rec0 = pysap.Image(data=np.abs(fourier_op.adj_op(kspace_obs)))
+# image_rec0.show()
+base_ssim = ssim(image_rec0, image)
+print('The SSIM from Adjoint is : ' + str(base_ssim))
+
+# Density Compensation adjoint
+image_rec1 = pysap.Image(data=np.abs(
+    fourier_op_density_comp.adj_op(kspace_obs))
```
Comment (zaccharieramzi) [inline examples/GPU_Examples/NonCartesian_gpuNUFFT_DensityCompensation.py:69]: specify that it's not really the adj op per se, since we precondition the k-space

File: mri/operators/fourier/utils.py [hunk 6]
```diff
@@ -308,3 +308,37 @@ def check_if_fourier_op_uses_sense(fourier_op):
...
+        implementation='gpuNUFFT',
+        osf=1,
+    )
+    density_comp = np.ones(kspace_loc.shape[0])
+    for _ in range(num_iterations):
+        density_comp = (
+                density_comp /
+                np.abs(grid_op.op(grid_op.adj_op(density_comp, True), True))
+        )
+    del grid_op
```
Comment (zaccharieramzi) [inline mri/operators/fourier/utils.py:343]: why do you need this? shouldn't it be garbage collected by Python?

File: mri/reconstructors/utils/extract_sensitivity_maps.py [hunk 7]
```diff
@@ -45,6 +46,9 @@ def extract_k_space_center_and_locations(data_values, samples_locations,
     is_fft: bool default False
         Checks if the incoming data is from FFT, in which case, masking
         can be done more directly
+    density_comp: np.ndarray default None
+        The density compensation for kspace data in case it exists and we
+        use density compensated adjoint for Smap estimation
     Returns
```
Comment (zaccharieramzi) [inline mri/reconstructors/utils/extract_sensitivity_maps.py:52]: Add the fact that density compensation might also be returned

File: mri/reconstructors/utils/extract_sensitivity_maps.py [hunk 8]
```diff
@@ -127,13 +143,19 @@ def get_Smaps(k_space, img_shape, samples, thresh,
             or len(thresh) != len(img_shape):
         raise NameError('The img_shape, max_samples, '
                         'min_samples and thresh must be of same length')
-    k_space, samples = \
+    k_space, *samples = \
```
Comment (zaccharieramzi) [inline mri/reconstructors/utils/extract_sensitivity_maps.py:146]: hmmm this is weird since `samples` might also have density compensation in it...
you can have `k_space, samples, *density_comp`, then do `if density_comp: density_comp = density_comp[0]`.

File: mri/test_local/test_fourier_adjoint_gpu.py [hunk 9]
```diff
@@ -35,11 +35,12 @@ def test_NUFFT_3D(self):
...
-                                                 shape=(self.N, self.N,
-                                                        self.N),
-                                                 implementation=platform,
-                                                 n_coils=num_channels)
+                fourier_op_dir = NonCartesianFFT(
+                    samples=_samples,
+                    shape=(self.N, self.N, self.N),
+                    implementation=platform,
+                    n_coils=num_channels
+                )
```
Comment (zaccharieramzi) [inline mri/test_local/test_fourier_adjoint_gpu.py:43]: not a great idea to have this change here. I understand what you tried to do but I think it's better left unchanged if no significant change is applied.

File: mri/test_local/test_fourier_adjoint_gpu.py [hunk 10]
```diff
@@ -23,7 +23,7 @@ class TestAdjointOperatorFourierTransformGPU(unittest.TestCase):
     def setUp(self):
         """ Set the number of iterations.
         """
-        self.N = 64
+        self.N = 128
```
Comment (zaccharieramzi) [inline mri/test_local/test_fourier_adjoint_gpu.py:26]: why this change?

File: mri/tests/test_fourier_adjoint.py [hunk 11]
```diff
@@ -113,18 +113,23 @@ def test_FFT(self):
         print(" FFT adjoint test passes")
 
     def test_NFFT_2D(self):
-        """Test the adjoint operator for the 2D non-Cartesian Fourier transform
+        """Test the adjoint operator for the 2D non-Cartesian Fourier
+        transform, with density compensator set to 1, to vet the code
+        path, the test is unchanged otherwise
```
Comment (zaccharieramzi) [inline mri/tests/test_fourier_adjoint.py:118]: I think I would be happier with 2 tests one without density compensation set and the other with it set to one

File: mri/tests/test_sensitivity_extraction.py [hunk 12]
```diff
@@ -63,23 +63,26 @@ def test_extract_k_space_center_3D(self):
...
+        data_thresholded, samples_thresholded, dc = \
             extract_k_space_center_and_locations(
                 data_values=np.reshape(Img, (self.num_channel,
                                              self.N * self.N)),
-                samples_locations=_samples,
+                samples_locations=samples,
                 thr=(thresh, thresh),
-                img_shape=(self.N, self.N))
+                img_shape=(self.N, self.N),
+                density_comp=np.ones(samples.shape[0])
```
Comment (zaccharieramzi) [inline mri/tests/test_sensitivity_extraction.py:84]: same here (i.e. with and without dc)

File: mri/operators/fourier/utils.py (hunk 6 above)
Comment (chaithyagr) [reply mri/operators/fourier/utils.py:343]: Well, I had a memory build up, so I had to add this.
But it turned out to be a bug in gpuNUFFT, which I have fixed and will make a side release for, I will remove this

File: mri/reconstructors/utils/extract_sensitivity_maps.py (hunk 8 above)
Comment (chaithyagr) [reply mri/reconstructors/utils/extract_sensitivity_maps.py:146]: Wait, does this work? I will check and update. We dont return anything sometimes

Comment (chaithyagr) [reply mri/reconstructors/utils/extract_sensitivity_maps.py:146]: Cool, this is good.

File: mri/test_local/test_fourier_adjoint_gpu.py (hunk 10 above)
Comment (chaithyagr) [reply mri/test_local/test_fourier_adjoint_gpu.py:26]: Debug change, will go back

File: mri/tests/test_fourier_adjoint.py (hunk 11 above)
Comment (chaithyagr) [reply mri/tests/test_fourier_adjoint.py:118]: Well, I really dont want to add on to tests as currently the time is nearly 40 minutes.
So the 3D works without density compensator while for 2D it works with density compensation. 
This way the code paths are vetted in both the cases. 
In the end the test is actually on the adjoint property more than density compensation itself. I havent really gotten a test for density compensation itself and I want to get one for it in future and that would be more meaningful perhaps to add?

File: mri/tests/test_sensitivity_extraction.py (hunk 12 above)
Comment (chaithyagr) [reply mri/tests/test_sensitivity_extraction.py:84]: Well here it makes sense, as we arent testing the other code paths at all. Adding this.

File: mri/reconstructors/utils/extract_sensitivity_maps.py (hunk 8 above)
Comment (zaccharieramzi) [reply mri/reconstructors/utils/extract_sensitivity_maps.py]: yes if you don't return density compensation you will have `density_comp == []` as you found out.

File: mri/operators/fourier/utils.py (hunk 6 above)
Comment (zaccharieramzi) [reply mri/operators/fourier/utils.py]: ok can you add a `#TODO` explaining that this will need change?

File: mri/tests/test_fourier_adjoint.py (hunk 11 above)
Comment (zaccharieramzi) [reply mri/tests/test_fourier_adjoint.py:118]: I see, well then one test here makes sense.

File: examples/GPU_Examples/NonCartesian_gpuNUFFT_DensityCompensation.py (hunk 1 above)
Comment (zaccharieramzi) [reply examples/GPU_Examples/NonCartesian_gpuNUFFT_DensityCompensation.py]: don't forget to port all the changes from this example to the other one (SENSE)

File: mri/operators/fourier/utils.py (hunk 6 above)
Comment (chaithyagr) [reply mri/operators/fourier/utils.py]: No No, this can be removed. This has no effect on memory, we just need to update gpuNUFFT which I already did. So this is kind of void at the moment, just was stale..

Comment (zaccharieramzi) [reply mri/operators/fourier/utils.py]: wait I didn't get that, you are already using the new gpuNUFFT release?

Comment (chaithyagr) [reply mri/operators/fourier/utils.py]: Well no, I will make a new release by end of day. But my point is that no matter whether we have the fix or not, this line does not help in anyway at all.. The underlying C++ and CUDA code had to free the memory, which wasnt, due to which we had a memory leak.

File: examples/GPU_Examples/NonCartesian_gpuNUFFT_DensityCompensation.py (hunk 1 above)
Comment (chaithyagr) [reply examples/GPU_Examples/NonCartesian_gpuNUFFT_DensityCompensation.py]: Yes on it

**Comment by chaithyagr** [comment]
All changes are in, @zaccharieramzi you can review again at convenience

File: mri/reconstructors/utils/extract_sensitivity_maps.py [hunk 13]
```diff
@@ -45,9 +46,14 @@ def extract_k_space_center_and_locations(data_values, samples_locations,
     is_fft: bool default False
         Checks if the incoming data is from FFT, in which case, masking
         can be done more directly
+    density_comp: np.ndarray default None
+        The density compensation for kspace data in case it exists and we
+        use density compensated adjoint for Smap estimation
     Returns
     -------
-    The extracted center of the k-space
+    The extracted center of the k-space. If the density compensators are
```
Comment (zaccharieramzi) [inline mri/reconstructors/utils/extract_sensitivity_maps.py:54]: Well since you are here you might as well correct the docs in full since we do not only return the center of the k-space but also the locations.

Also don't forget to indicate the dtypes.

File: mri/tests/test_sensitivity_extraction.py (hunk 12 above)
Comment (zaccharieramzi) [reply mri/tests/test_sensitivity_extraction.py:84]: you didn't do it right? or am I missing sthg?

File: mri/reconstructors/utils/extract_sensitivity_maps.py (hunk 13 above)
Comment (chaithyagr) [reply mri/reconstructors/utils/extract_sensitivity_maps.py:54]: Sorry the dtypes? I have changed the sentence to: 
```
The extracted center of the k-space, i.e. both the kspace locations and kspace values.
```

Arent the return dtypes same as input?

Comment (zaccharieramzi) [reply mri/reconstructors/utils/extract_sensitivity_maps.py:54]: Well surely, but it's still good to remind that

File: mri/tests/test_sensitivity_extraction.py (hunk 12 above)
Comment (chaithyagr) [reply mri/tests/test_sensitivity_extraction.py:84]: I did add it, it is in the same test. Below you can see that I added another Smap estimation

File: mri/tests/test_sensitivity_extraction.py [hunk 14]
```diff
@@ -114,30 +117,42 @@ def test_sensitivity_extraction_2D(self):
...
             k_space=F_img,
             img_shape=(self.N, self.N),
-            samples=_samples,
+            samples=samples,
             thresh=(0.4, 0.4),
             mode='gridding',
             min_samples=(-0.5, -0.5),
             max_samples=(0.5, 0.5),
             n_cpu=1)
+        Smaps_NFFT_dc, SOS_Smaps_dc = get_Smaps(
```
Comment (chaithyagr) [inline mri/tests/test_sensitivity_extraction.py:136]: This is the other one
